### ISSUE TITLE
JsonQueryResultsReader is dependant on ordering of keys in JSON result

### ISSUE THREAD

**Issue by epoz** [oxigraph/oxigraph#429]
**Describe the bug**
If a sparql query result read by the JsonQueryResultsReader contains the `head` key _after_ the `results` key, like so:

```
{'results': {'bindings': [
  {'s': {'type': 'uri', 'value': 'urn:foo'},
    'p': {'type': 'uri', 'value': 'urn:bar'},
    'o': {'type': 'uri', 'value': 'urn:baz'}}
 ]},
 'head': {'vars': ['p', 'o', 's']}}
```

JsonQueryResultsReader gives an error message. The [spec](https://www.w3.org/TR/sparql11-results-json/) does not explicitly mention ordering of keys, but in the examples and most implementations the head comes first.

When the results are generated by systems where you have no control of the ordering of the JSON, (like Python rdlib rdflib.plugins.sparql.results.jsonresults.JSONResultSerializer) - the `head` might come after the results.

(see here: https://github.com/oxigraph/oxigraph/blob/be3b009f5df427c0bc80cf8883b66a96644fea63/lib/sparesults/src/json.rs#L186)

**To Reproduce**
1. Which version of Oxigraph are you using? On which platform?

Identified the problem using:
pyoxigraph==0.3.8
oxrdflib==0.3.0

Using Python on the command line, 
3. A command line or a code snippet that triggers the bug.

Using a service query in the Python rdflib from the REPL like:
```
SELECT ?obj WHERE {
  SERVICE <http://localhost:9000/sparql> {
    ?sub ?pred ?obj .
 }
}
```

- on port 9000 is a HTTP server that can return either,
`'{"results": {"bindings": [{"s": {"type": "uri", "value": "urn:foo"}, "p": {"type": "uri", "value": "urn:bar"}, "o": {"type": "uri", "value": "urn:baz"}}]}, "head": {"vars": ["p", "o", "s"]}}`
or
`{"head": {"vars": ["p", "o", "s"]}, "results": {"bindings": [{"s": {"type": "uri", "value": "urn:foo"}, "p": {"type": "uri", "value": "urn:bar"}, "o": {"type": "uri", "value": "urn:baz"}}]}}` on the `/sparql` path as a POST response with content-type 'application/sparql-results+json'

**Comment by Tpt** [comment]
@epoz Thank you for trying Oxigraph and sorry for this bug.

The SPARQL JSON parser is implemented in a streaming way and relies on the list of the variables in the header to initialize data structures (to save some space and compute time bindings set are encoded using arrays and not hash map in Oxigraph). But results with the `head` key last should definitely be supported. I believe I will add support of this case, maybe dropping the "streaming" part of the parser.

**Comment by Tpt** [comment]
Related MR: #430

**Comment by epoz** [comment]
No rush from this side, thanks for your efforts.
After figuring out what the problem was, I could make a work-around in our code.
(re-doing the Python rdflib for the results serialization, but forcing the head key to be first)

But good to know that it will also be fine up-stream eventually.

**Comment by Tpt** [comment]
Now released in 0.3.14.
